src/Bicep.Core/Semantics/ImportedSymbol.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics.Metadata;
+using Bicep.Core.Registry;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem;
+
+namespace Bicep.Core.Semantics;
+
+public class ImportedSymbol : DeclaredSymbol
+{
+    private readonly Lazy<ISemanticModel?> sourceModelLazy;
+
+    public ImportedSymbol(ISymbolContext context, ImportedSymbolsListItemSyntax declaringSyntax, CompileTimeImportDeclarationSyntax enclosingDeclartion)
+        : base(context, declaringSyntax.Name.IdentifierName, declaringSyntax, declaringSyntax.Name)
+    {
+        EnclosingDeclaration = enclosingDeclartion;
+
+        sourceModelLazy = new(() =>
+        {
+            SemanticModelHelper.TryGetSemanticModelForForeignTemplateReference(Context.Compilation.SourceFileGrouping,
```

Comment:
I was curious about the decision to build in caching here. Looking at the implementation for this method - it looks like `UriResultByArtifactReference` is of type `ImmutableDictionary<ISourceFile, ImmutableDictionary<IArtifactReferenceSyntax, UriResolutionResult>>` - e.g. we have to loop over values to search each file for artifact references.

Do you know why that's the case? IIRC it used to be a flat dictionary of `ImmutableDictionary<IArtifactReferenceSyntax, UriResolutionResult>`

Reply:
It looks like [I introduced the tiered dictionary a while back](https://github.com/Azure/bicep/pull/8301) so that registry aliases could be resolved against the correct `bicepconfig.json` file, but it'd be worth keeping a flattened copy in `SourceFileGroupingBuilder` to make lookups cheaper.